keps/sig-api-machinery/3037-client-go-alternative-services/README.md

```diff
@@ -0,0 +1,590 @@
+
+## Motivation
+
+Currently, when deploying high available clusters, clients can only benefit of it by using external load balancers or mechanisms that balance the connection to the corresponding control-plane.
+This also has the "chicken and egg" problem on some components like kube-proxy and kubenet and "in-cluster" clients, that require `Services` to load balance the connection.
+
+Ref: https://github.com/kubernetes/kubernetes/issues/18174
+
+### Goals
+
+- Allow users to don't require external load balancers between cluster and control-plane for HA setups.
+- Allow components to bootstrap without any dependency on external load balancers or `Services` implementations.
+- Increase the resilience of the cluster.
+
+### Non-Goals
+
+- The goal is to offer the opt-in functionality, any component change to use it is out of scope.
+- Provide a lot of nines of availability.
+- Work outside of the cluster network.
+- Provide a client-side load balancing solution to distribute the load between the available API servers.
+
+## Proposal
+
+The proposal is to extend client-go to allow to connect to multiple control-plane nodes, and for API servers to announce the list of available servers to the clients.
+
+### User Stories (Optional)
+
+#### Story 1
+
+As a Kubernetes admin I want to deploy my cluster in HA without having to use external load balancers for the cluster communication.
+
+#### Story 2
+
+As a Kubernetes developer I'd like to be able to develop HA applications and controllers without requiring external components.
+
+#### Story 3
+
+As a Kubernetes architect I'd like the cluster components to be completely autonomous.
+
+### Notes/Constraints/Caveats (Optional)
+
+Network load balancing is complicated, there are many edge cases and there is no universal solution, the requirements also change and evolve with the time (see HTTP1.1 vs HTTP2).
+
+Client-go is the library used by all the cluster components to communicate with the API servers, any change or regression there can have a great impact on the project.
+
+The implementation of this feature should not modify any of current behaviors and be completely pluggable, so users can opt-in safely and the project can be sure there will not create any regression.
+
+### Risks and Mitigations
+
+The feature will be opt-in, users can disable it and keep having the same behavior.
+
+To avoid possible security issues, this feature will require HTTP2 and HTTPS in order to work.
+
+The alternative services IPs exposed are the same published in the Endpoints for the `kubernetes.default` services, the Alt-Svc headers will not be present on requests
+that are not authenticated.
+
+## Design Details
+
+There are two problems we have to solve:
+
+- get a list of available control-plane nodes
+- client capability to be able to connect to multiple control-planes
+
+### List of available control-plane nodes
+
+This problem can be solved in two different ways:
```

Comment:
I expected to see DNS with multiple A records in this list, why not that?

Reply:
DNS has a dependency problem, same chicken an egg that Services., however, there is an option to have an in-memory dns, I've implemented it in https://github.com/kubernetes/enhancements/pull/3034/files#diff-9a0675306e36179b91330951d5555c276fba6b0ec053d50e3c143d449d1f8543R582 as alternative.

The main problem is you have to override the dialer